contrib/epee/include/net/local_ip.h

```diff
@@ -27,10 +27,33 @@
 
 #pragma once
 
+#include <string>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/asio/ip/address_v6.hpp>
+
 namespace epee
 {
   namespace net_utils
   {
+
+    inline
+    bool is_ipv6_local(const std::string& ip)
+    {
+      // ipv6 local addresses start with fc00/7 -- (fcXX or fdXX)
+      const std::string first_two = ip.substr(0, 2);
+      static const std::string match_str1 = "fc";
+      static const std::string match_str2 = "fd";
+
+      return (boost::iequals(first_two, match_str1) || boost::iequals(first_two, match_str2));
```

Comment:
I'd rather this used the boost from_string + is_link_local, since it turns out those exist.

Reply:
I looked at that as well but I don't think is_link_local does exactly what we want.

https://github.com/boostorg/asio/blob/develop/include/boost/asio/ip/impl/address_v6.ipp#L168

for both is_link_local and is_site_local it's looking if the address is link-local, but I don't think that's what we care about.  I could be mistaken here, but I presume we care about unique-local-addresses in IPv6, i.e. the equivalent of 10.0.0.0/8 or 192.168/16

If link-local is indeed what we care about it can certainly be changed.

Reply:
Looking at the docs, is_site_local is what we want here.

Reply:
Actually, not sure. I see a claim that site-local allows talking to other machines on the LAN, which is what we want (if there is a "only"). So I don't get the difference between link and site then.

Reply:
Well, so I'm not sure in practice which would matter.  Could even check for both.

I just know that the wikipedia article compared link-local in IPv6 as the equivalent of 169.254.0.0/16 in IPv4.  It also mentioned something the spec requiring every interface to have a link-local address, though I did not quite understand to what end.

I guess maybe just go with both?

Reply:
I think we want site local. The use of it is to detect when a node is not a loopback node, but is still in the LAN, to default to trusted. Site local seems to encompass all the non routable IPv4 address ranges that are meant for LAN only use.

Reply:
So looking into it on IANA's site, site-local is deprecated as of a 2004 RFC.

RFC 4193 states that Unique Local Unicast addresses are intended for use within a "site" and not intended to be routed on the internet.

RFC 4291 indicates that routers are not intended to forward packets from link-local addresses (fe80::/10) but I guess it could sorta still be used similarly to loopback, though that would be very strange indeed.

I shall add is_site_local to the function, but the current check for unique local unicast should remain.

Reply:
OK. Then we'll just wait for someone with an IPv6 LAN to tell us whether it works as expected or not.